File: bst.erl
%%%--------------------------------------------------------------------------------------------------%%%
%%% @author Shahzad Bhatti
%%% @doc Binary Search Tree
%%%     erlc bst.erl
%%%     erl -noshell -s bst test_bst_search -s init stop
%%%--------------------------------------------------------------------------------------------------%%%

-module(bst).
-export([bst_search/2, test_bst_search/0]).

-record(node, {data, left, right}).


bst_search(undefined, _Key) ->
   undefined;
bst_search(Node, Key) when Node#node.data == Key ->
   Node;
bst_search(Node, Key) when Node#node.data > Key ->
   bst_search(Node#node.left, Key);
bst_search(Node, Key) ->
   bst_search(Node#node.right, Key).

make_node(Data) -> 
   #node{data = Data}.
make_node(Data, Left, Right) when is_integer(Left), is_integer(Right) -> 
   #node{data = Data, left = make_node(Left), right = make_node(Right)};
make_node(Data, Left, Right) when is_integer(Left), not is_integer(Right) -> 
   #node{data = Data, left = make_node(Left), right = Right};
make_node(Data, Left, Right) when not is_integer(Left), is_integer(Right) -> 
   #node{data = Data, left = Left, right = make_node(Right)};
make_node(Data, Left, Right) ->
   #node{data = Data, left = Left, right = Right}.


%%%%
%         8
%    3       10
% 1    6        14
%    4        13
%%%%
create_test_data() ->
   N6 = make_node(6, 4, 7),   
   N14 = make_node(14, 13, undefined),
   N3 = make_node(3, 1, N6),
   N10 = make_node(10, undefined, N14),
   make_node(8, N3, N10).


print_node(undefined, _Indent) ->
   undefined;
print_node(Node, _Indent) when Node#node.data == undefined ->
   undefined;
print_node(Node, Indent) ->
    Spaces = string:chars($ , Indent),
    io:fwrite("~s~w\n", [Spaces, Node#node.data]),
    print_node(Node#node.left, Indent+1),
    print_node(Node#node.right, Indent+1).
   
test_bst_search() ->
    Root = create_test_data(),
    %print_node(Root, 0),
    Node1 = bst_search(Root, 6),
    6 = Node1#node.data,
    Node2 = bst_search(Root, 14),
    14 = Node2#node.data,
    undefined = bst_search(Root, 61).

File: BST.java
public class BST {
   private class Node {
      private int val;
      private Node left;
      private Node right;
      public Node(int val) {
         this.val = val;
      }
      @Override
      public String toString() {
         return "" + val;
      }
   }

   private Node root;

   public void insert(int val) {
      root = insert(root, val);
   }

   private Node insert(Node p, int val) {
      if (p == null) return new Node(val);
      if (val == p.val) {
         return p;
      } else if (val < p.val) {
         p.left = insert(p.left, val);
      } else {
         p.right = insert(p.right, val);
      }
      return p;
   }

   public boolean search(int val) {
      return search(root, val);
   }

   private boolean search(Node p, int val) {
      if (p == null) {
         return false;
      } else if (val == p.val) {
         return true;
      } else if (val < p.val) {
         return search(p.left, val);
      } else {
         return search(p.right, val);
      }
   }

   public void delete(int val) {
      root = delete(root, val);
   }
   private Node delete(Node p, int val) {
      if (p == null)  {
         return p;
      } else if (val < p.val) {
         p.left = delete (p.left, val);
      } else if (val > p.val) {
         p.right = delete (p.right, val);
      } else {
         if (p.left == null) {
            return p.right;
         } else if (p.right == null) {
            return p.left;
         } else {
            // get data from the rightmost node in the left subtree
            p.val = retrieveData(p.left);
            // delete the rightmost node in the left subtree
            p.left =  delete(p.left, p.val) ;
         }
      }
      return p;
   }
   private int retrieveData(Node p) {
      while (p.right != null) p = p.right;
      return p.val;
   }

   public void preOrderTraversal() {
      preOrderHelper(root);
   }
   private void preOrderHelper(Node r) {
      if (r != null) {
         System.out.print(r+" ");
         preOrderHelper(r.left);
         preOrderHelper(r.right);
      }
   }

   public void inOrderTraversal() {
      inOrderHelper(root);
   }
   private void inOrderHelper(Node r) {
      if (r != null) {
         inOrderHelper(r.left);
         System.out.print(r+" ");
         inOrderHelper(r.right);
      }
   }

   public int height() {
      return height(root);
   }
   private int height(Node p) {
      if (p == null) return -1;
      else
         return 1 + Math.max( height(p.left), height(p.right));
   }

   public int countLeaves() {
      return countLeaves(root);
   }
   private int countLeaves(Node p) {
      if(p == null) {
         return 0;
      } else if(p.left == null && p.right == null) {
         return 1;
      } else {
         return countLeaves(p.left) + countLeaves(p.right);
      }
   }



   public static void main(String[] args) {
      int[] a = {1,5,2,7,4};
      BST bst = new BST();
      for(int n : a) bst.insert(n);

      bst.preOrderTraversal();
      System.out.println();
   }
}
